File: langstons_ant/samjones/lib/ant.rb
require 'grid_navigable'

class Ant
	include GridNavigable
	@@directions = [
		:north,
		:east,
		:south,
		:west
	]

	@@turn_direction = {
		left: -1,
		right: +1
	}

	def initialize(x, y, opts = {})
		at(x, y)
		@facing = opts[:facing] || :north
		@printer = opts[:printer] || :printer
	end

	def turn!(turn)
			increment = @@turn_direction[turn]
			current_pos = @@directions.find_index(@facing)
			@facing = @@directions[(current_pos + increment) % @@directions.size]
		self
	end

	def move!
		move_direction!(@facing)
		self
	end

	def print
		@printer.print(x, y, @facing) if @printer
		self
	end
end

File: langstons_ant/nathan/src/langton/ant.clj
(ns langton.ant
  (:require [langton.rules :as rules]))

(def directions [:north :west :south :east])

(defn- next-pos [[x y] faces]
  (condp = faces
    :north [     x  (inc y)]
    :west  [(dec x)      y ]
    :south [     x  (dec y)]
    :east  [(inc x)      y ]))

(defn create [& {:keys [pos faces]
                 :or {pos [0 0] faces :west}}]
  {:pos pos :faces faces})

(defn turn [{:keys [faces] :as ant} left-or-right]
  {:pre [(left-or-right #{:left :right})]}
  (let [current-index (.indexOf directions faces)
        inc-or-dec (left-or-right {:left inc :right dec})
        next-index (mod (inc-or-dec current-index) (count directions))]
    (assoc ant :faces (get directions next-index))))

(defn step [{:keys [pos faces] :as ant}]
  (assoc ant :pos (next-pos pos faces)))
